### PR TITLE
Wallet connect - block unsupported

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Part of and pipe into #597

Wallet connect - block unsupported wallets

screenshot_2021-05-17_13-15-06
screenshot_2021-05-17_13-14-22
screenshot_2021-05-17_13-13-21

Those include:
- Any SmartContract wallet
- Crypto.com Defi Wallet
- TokenPocket

Regarding TokenPocket, they are working on a fix https://github.com/TP-Lab/tp-android/issues/16. 
We might be able to remove this block soon.

**Bonus**:
screenshot_2021-05-17_13-16-18
screenshot_2021-05-17_13-16-10

- Added wallet images to details modal (it was disabled)
- Show connected wallet name, if available
- Show connected wallet icon, if available

# Testing

1. Connect an EOA (Metamask, Nifty) or injected (dApp browser such as imToken, TokenPocket) wallet
- [ ] Swap button should be enabled

1. Using WalletConnect, connect a SmartContract wallet (like Gnosis Safe, Argent, Pillar)
- [ ] Swap button should be disabled
- [ ] The icon next to the account address should be a :warning: and show a tooltip on hover
- [ ] On account details, the icon next to the account address should be a :warning: and show a tooltip on hover

----

## Note

This PR is already too big.
Will add an FAQ entry in a follow up.

### CODE REVIEW

File: src/custom/components/AccountDetails/index.tsx [hunk 1]
```diff
@@ -38,22 +38,44 @@ import {
...
 } from './AccountDetailsMod'
+import { ConnectedWalletInfo, useWalletInfo } from 'hooks/useWalletInfo'
+import { MouseoverTooltip } from 'components/Tooltip/TooltipMod'
 
 type AbstractConnector = Pick<ReturnType<typeof useActiveWeb3React>, 'connector'>['connector']
 
-export function formatConnectorName(connector?: AbstractConnector) {
+export function formatConnectorName(connector?: AbstractConnector, walletInfo?: ConnectedWalletInfo) {
   const { ethereum } = window
+  const { walletName } = walletInfo || {}
```
Comment (W3stside) [inline src/custom/components/AccountDetails/index.tsx:48]: Nit: could also set the default in the param and remove the undefined type

File: src/custom/components/AccountDetails/index.tsx [hunk 2]
```diff
@@ -38,22 +38,44 @@ import {
...
-    .filter(
-      k => SUPPORTED_WALLETS[k].connector === connector && (connector !== injected || isMetaMask === (k === 'METAMASK'))
-    )
-    .map(k => SUPPORTED_WALLETS[k].name)[0]
+  const name =
+    walletName ||
+    Object.keys(SUPPORTED_WALLETS)
+      .filter(
+        k =>
+          SUPPORTED_WALLETS[k].connector === connector && (connector !== injected || isMetaMask === (k === 'METAMASK'))
```
Comment (W3stside) [inline src/custom/components/AccountDetails/index.tsx:55]: I know you didn't write this but this line sucks to read lol

**Comment by W3stside** [comment]
Details: 
- can connect via metamask mobile WC, see logs, shows supported and WalletConnect ✅ 
- no app breaking on connecting ✅ 
Issues:
- no loading balances ❌ 
- error message:
image

File: src/custom/components/AccountDetails/index.tsx (hunk 2 above)
Comment (alfetopito) [reply src/custom/components/AccountDetails/index.tsx:55]: Agreed 😮‍💨

Will see what I can do

Comment (alfetopito) [reply src/custom/components/AccountDetails/index.tsx]: Gave it a try on https://github.com/gnosis/gp-swap-ui/pull/633/commits/0d2b9e4135ab8d8ef113a361f671c2326bc0e225

Not very happy with it yet, but I've refactored some of the ugliness away

**Comment by alfetopito** [comment]
> Details:
> 
>     * can connect via metamask mobile WC, see logs, shows supported and WalletConnect white_check_mark
> 
>     * no app breaking on connecting white_check_mark
>       Issues:
> 
>     * no loading balances x
> 
>     * error message:
>       image

Hmmm, it worked fine for me :/
How long did you wait for?
Have you tried to refresh the page?

File: .eslintrc.json [hunk 3]
```diff
@@ -21,11 +19,12 @@
     "plugin:@typescript-eslint/recommended",
     "plugin:react-hooks/recommended",
     "prettier/@typescript-eslint",
-    "plugin:prettier/recommended"
+    "plugin:prettier/recommended",
+    "plugin:jsx-a11y/recommended"
```
Comment (anxolin) [inline .eslintrc.json:23]: What is this change? (is this related somehow to this PR?

File: src/custom/components/AccountDetails/index.tsx [hunk 4]
```diff
@@ -38,22 +38,56 @@ import {
...
-      k => SUPPORTED_WALLETS[k].connector === connector && (connector !== injected || isMetaMask === (k === 'METAMASK'))
-    )
-    .map(k => SUPPORTED_WALLETS[k].name)[0]
-  return <WalletName>Connected with {name}</WalletName>
+
+  const walletTuple = Object.entries(SUPPORTED_WALLETS).filter(
+    ([walletType, { connector: walletConnector }]) =>
+      walletConnector === connector && (connector !== injected || isMetaMask === (walletType === 'METAMASK'))
+  )
+  return walletTuple[0]?.[1]?.name || 'Unknown wallet'
```
Comment (anxolin) [inline src/custom/components/AccountDetails/index.tsx:54]: I get a bit lost. I understand the signature, but the implementation is a bit hard to follow. 
Maybe at least some comment. 

Also, I thought we were getting all the info from `useWalletInfo` are these logics something we want in the hook? (so components don't need to know about strange algorithms to get our best guess in the walletName?

File: src/custom/components/AccountDetails/index.tsx [hunk 5]
```diff
@@ -38,22 +38,56 @@ import {
...
+}
+
+export function formatConnectorName(connector?: AbstractConnector, walletInfo?: ConnectedWalletInfo) {
+  const name = walletInfo?.walletName || getWalletName(connector)
+  // In case the wallet is connected via WalletConnect and has wallet name set, add the suffix to be clear
+  // This to avoid confusion for instance when using Metamask mobile
+  // When name is not set, it defaults to WalletConnect already
+  const walletConnectSuffix =
+    walletInfo?.provider === WalletProvider.WALLET_CONNECT && walletInfo?.walletName ? ' (via WalletConnect)' : ''
```
Comment (anxolin) [inline src/custom/components/AccountDetails/index.tsx:63]: Why do you also chek for `walletInfo?.walletName`? 

is not enough `walletInfo?.provider === WalletProvider.WALLET_CONNECT`

File: src/custom/components/AccountDetails/index.tsx [hunk 6]
```diff
@@ -38,22 +38,56 @@ import {
...
+
+  return (
+    <WalletName>
+      Connected with {name}
+      {walletConnectSuffix}
+    </WalletName>
```
Comment (anxolin) [inline src/custom/components/AccountDetails/index.tsx:69]: This is very cool! 👌

**Review by anxolin — APPROVED** [review]
Very good PR! ❤️

File: src/custom/components/AccountDetails/index.tsx (hunk 4 above)
Comment (alfetopito) [reply src/custom/components/AccountDetails/index.tsx:54]: This is my attempt to make the original uniswap code more readable. See David's comment https://github.com/gnosis/gp-swap-ui/pull/633#discussion_r634122893
Failed miserably, apparently.

File: .eslintrc.json (hunk 3 above)
Comment (alfetopito) [reply .eslintrc.json:23]: Yes.
The emoji rule did not like that I am setting the role + aria-label on something other than a span (it's div Styled component).
I'm disabling that rule in the code block where this happens, but vscode doesn't know about the rule without this change, and thinks the code is broken otherwise.

File: src/custom/components/AccountDetails/index.tsx (hunk 5 above)
Comment (alfetopito) [reply src/custom/components/AccountDetails/index.tsx:63]: In my defense, this one does have a comment :)

>  // When name is not set, it defaults to WalletConnect already

Avoiding a wallet with name: WalletConnect (via WalletConnect)

**Comment by alfetopito** [comment]
@W3stside Tested what you mentioned.
1. Connected with MM mobile.
2. Balance loads, all good.
3. Switch networks in the app
4. Not reflected in the UI at all
    Had to disconnect and connect again in the other network to make it work.

Will create another issue for this.
